Server-to-server OAuth #387 (#403)
ncstate-delta/moodle-mod_zoom: README.md

> 

Reviewer: The Prerequisites line above this should be a second-level heading.
```
## Prerequisites
```

This plugin is designed for Educational or Business Zoom accounts.

> To connect to the Zoom APIs this plugin requires an account level app to be

Reviewer: ```suggestion
To connect to the Zoom APIs, this plugin requires an account-level app to be
```

created.

> ## Server-to-Server OAuth

Reviewer: ```suggestion
### Server-to-Server OAuth
```

Author: Done

> To create an account-level server-to-server oauth app, the Server-to-server OAuth app
> Permission is required.

Reviewer: ```suggestion
To [create an account-level Server-to-Server OAuth app](https://marketplace.zoom.us/docs/guides/build/server-to-server-oauth-app), the `Server-to-server OAuth app`
permission is required.
```

> See https://marketplace.zoom.us/docs/guides/build/server-to-server-oauth-app.

Reviewer: ```suggestion
```

Author: Done

> You will need to create a Server-to-Server OAuth app and that will generate
> the client ID, client secret, and account ID.

Reviewer: ```suggestion
The Server-to-Server OAuth app will generate a client ID, client secret and account ID.
```

Author: Done

> The following scopes are used by this plugin:
> 
> - account:read:admin
> - dashboard_meetings:read:admin
> - dashboard_webinars:read:admin
> - meeting:read:admin
> - meeting:write:admin
> - recording:read:admin
> - report:read:admin
> - tracking_fields:read:admin
> - user:read:admin
> - user:write:admin
> - webinar:read:admin
> - webinar:write:admin
> 
> Scopes outlined by plugin settings:
> 
> - Create/edit/delete meetings/webinars
>     - account:read:admin
>     - meeting:read:admin
>     - meeting:write:admin
>     - webinar:read:admin
>     - webinar:write:admin
>     - user:read:admin
>     - report:read:admin (Pro user and up)
>     - dashboard_meetings:read:admin (Business accounts and higher)
>     - dashboard_webinars:read:admin  (Business accounts and higher)
> - Number of licenses (zoom | licensesnumber)
>     - user:read:admin
> - Redefine licenses (zoom | utmost)
>      - user:read:admin
>     - user:write:admin
> - Recycle license upon join (zoom | recycleonjoin)
>     - user:read:admin
>     - user:write:admin
> - Allow recordings to be viewed (zoom | viewrecordings)
>     - recording:read:admin
> - Webinar (zoom | showwebinars)
>     - user:read:admin
>     - webinar:write:admin
>     - webinar:read:admin
>     - dashboard_webinars:read:admin
> - Webinar by default (zoom | webinardefault)
>     - user:read:admin
>     - webinar:write:admin
>     - webinar:read:admin
>     - dashboard_webinars:read:admin
> - Scheduling privilege (zoom | showschedulingprivilege)
>     - user:read:admin
> - Alternative Hosts (zoom | showalternativehosts (Show alternative hosts option as user picker with autocompletion))
>     - user:read:admin
> - Meeting capacity warning (zoom | showcapacitywarning)
>     - user:read:admin
> - Download iCal (zoom | showdownloadical)
>     - meeting:read:admin 
> - Automatic recording (zoom | recordingoption (Use default Zoom user settings))
>     - user:read:admin
> - Tracking fields (zoom | defaulttrackingfields)
>     - tracking_fields:read:admin

Reviewer: I had this idea to try and make it a bit shorter and more "à la carte". Please take a look and make sure I didn't oversimplify it to the point of losing important distinctions.
```suggestion
At a minimum, the following scopes are required by this plugin:

- meeting:read:admin (Read meeting details)
- meeting:write:admin (Create/Update meetings)
- user:read:admin (Read user details)

Additional scopes are required for certain functionality:

- Account-level passcode / encryption policies
    - account:read:admin
- Reports for meetings / webinars
    - dashboard_meetings:read:admin (Business accounts and higher)
    - dashboard_webinars:read:admin  (Business accounts and higher)
    - report:read:admin (Pro user and up)
- Allow recordings to be viewed (zoom | viewrecordings)
    - recording:read:admin
- Tracking fields (zoom | defaulttrackingfields)
    - tracking_fields:read:admin
- Recycle licenses (zoom | utmost), (zoom | recycleonjoin)
    - user:write:admin
- Webinars (zoom | showwebinars), (zoom | webinardefault)
    - webinar:read:admin
    - webinar:write:admin
```

Author: Done

> ## JWT

Reviewer: ```suggestion
### JWT
```

Author: Done

> JWT will be deprecated in June 2023. To create an account-level JWT app the Developer Role Permission is

Author: Done

See https://marketplace.zoom.us/docs/guides/build/jwt-app. You will need to
create a JWT app and that will generate the API key and secret.

> JWT will be deprecated in June 2023; see https://marketplace.zoom.us/docs/guides/build/server-to-server-oauth-app.

Author: Done

You will need to create a Server-to-Server OAuth app and that will generate
the client ID, client secret, and account ID.

The following scopes are used by this plugin:

- account:read:admin
- dashboard_meetings:read:admin
- dashboard_webinars:read:admin
- meeting:read
- meeting:read:admin
- meeting:write
- meeting:write:admin
- recording:read
- recording:read:admin
- report:read:admin
- tracking_fields:read:admin
- user_info:read
- user:read
- user:read:admin
- user:write:admin
- webinar:read
- webinar:read:admin
- webinar:write

> - webinar:write:admin

Reviewer: We probably don't use the non `:admin` scopes. We may want to specify - at least generally - which pieces of functionality use each scope. Especially for the more-sensitive ones.

Author: I've added scopes outlined by plugin setting to the README now.

If you get "Access token is expired" errors, make sure the date/time on your
server is properly synchronized with the time servers.

> After JWT deprecation in June 2023, you need to set the following  settings to enable the plugin:

Author: Done

## Installation

1. Install plugin to mod/zoom. More details at https://docs.moodle.org/39/en/Installing_plugins#Installing_a_plugin
2. Once you install the plugin you need to set the following settings to enable the plugin:

- Zoom client ID (mod_zoom | clientid)
- Zoom client secret (mod_zoom | clientsecret)
- Zoom account ID (mod_zoom | accountid)

JWT will be deprecated in June 2023. For a JWT app, you need to set the following settings to enable the plugin:

- Zoom API key (mod_zoom | apikey)

> - Zoom API secret (mod_zoom | apisecret)

Reviewer: The zoomurl setting (one line below this) is not tied to the JWT API keys. It's mostly just to give people links to Zoom.

Author: Done
